iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/compaction/execute/utils/executor/batch/utils/AlignedSeriesBatchCompactionUtils.java

```diff
@@ -173,4 +172,75 @@ public static ModifiedStatus checkIsModified(
     }
     return status;
   }
+
+  public static class BatchColumnSelection {
+    private final List<IMeasurementSchema> schemaList;
+    private final LinkedList<Integer> typeSortedColumnIndexList;
+    private final LinkedList<Integer> largeTypeSortedColumnIndexList;
+    private final int batchSize;
+    private int selectedColumnNum;
+
+    private List<Integer> columnIndexList;
+    private List<IMeasurementSchema> currentSelectedColumnSchemaList;
+
+    public BatchColumnSelection(List<IMeasurementSchema> valueSchemas, int batchSize) {
+      this.schemaList = valueSchemas;
+      this.typeSortedColumnIndexList = new LinkedList<>();
+      this.largeTypeSortedColumnIndexList = new LinkedList<>();
+      for (int i = 0; i < valueSchemas.size(); i++) {
+        IMeasurementSchema schema = valueSchemas.get(i);
+        if (isLargeDataType(schema.getTypeInByte())) {
+          this.largeTypeSortedColumnIndexList.addLast(i);
+        } else {
+          this.typeSortedColumnIndexList.addLast(i);
+        }
+      }
+      this.batchSize = batchSize;
+      this.selectedColumnNum = 0;
+    }
+
+    public boolean hasNext() {
+      return selectedColumnNum < schemaList.size();
+    }
+
+    public void next() {
+      if (!hasNext()) {
+        throw new NoSuchElementException();
+      }
+      selectColumnBatchToCompact();
+    }
+
+    public List<Integer> getSelectedColumnIndexList() {
+      return this.columnIndexList;
+    }
+
+    public List<IMeasurementSchema> getCurrentSelectedColumnSchemaList() {
+      return this.currentSelectedColumnSchemaList;
+    }
+
+    private void selectColumnBatchToCompact() {
+      // TODO: select batch by allocated memory and chunk size to perform more strict memory control
+      this.columnIndexList = new ArrayList<>(batchSize);
+      this.currentSelectedColumnSchemaList = new ArrayList<>(batchSize);
+      while (!typeSortedColumnIndexList.isEmpty() || !largeTypeSortedColumnIndexList.isEmpty()) {
+        LinkedList<Integer> sourceTypeSortedList =
+            largeTypeSortedColumnIndexList.isEmpty()
+                ? typeSortedColumnIndexList
+                : largeTypeSortedColumnIndexList;
+        Integer columnIndex = sourceTypeSortedList.removeFirst();
+        this.columnIndexList.add(columnIndex);
+        this.currentSelectedColumnSchemaList.add(this.schemaList.get(columnIndex));
+        selectedColumnNum++;
+        if (this.columnIndexList.size() >= batchSize) {
+          break;
+        }
+      }
+    }
```

Comment:
A queue seems to be more suitable for this job than LinkedList. Or you may remove a sublist from the list to reduce iteration.